test_cases/test_easygui.py

```diff
@@ -0,0 +1,124 @@
+import sys
+sys.path.append('..')
+import easygui
+
+import inspect
+import os
+import time
+import threading
+
+import pytest
+from pynput.keyboard import Key, Controller
+
+KEYBOARD = Controller()
+FOLDER_OF_THIS_FILE = os.path.dirname(os.path.abspath(__file__))
+GUI_WAIT = 0.6 # if tests start failing, maybe try bumping this up a bit (though that'll slow the tests down)
+
+
+"""
+NOTE: You will often see this code in this test:
+
+    print('Line', inspect.currentframe().f_lineno);
+
+This is because due to the GUI nature of these tests, if something messes up
+and PyAutoGUI is unable to click on the message box, this program will get
+held up. By printing out the line number, you will at least be able to see
+which line displayed the message box that is held up.
+
+This is a bit unorthodox, and I'm welcome to other suggestions about how to
+deal with this possible scenario.
+"""
+
+class KeyPresses(threading.Thread):
+    def __init__(self, keyPresses):
+        super(KeyPresses, self).__init__()
+        self.keyPresses = keyPresses
+
+    def run(self):
+        time.sleep(GUI_WAIT)
+        KEYBOARD.type(self.keyPresses)
+
+
+def test_test_images_exist():
+    assert os.path.exists(os.path.join(FOLDER_OF_THIS_FILE, 'pi.jpg'))
+    assert os.path.exists(os.path.join(FOLDER_OF_THIS_FILE, 'result.png'))
+
+def test_msgbox():
+    # Test hitting space to click OK.
+    t = KeyPresses(' ')
+    t.start()
+    print('Line', inspect.currentframe().f_lineno)
+    assert easygui.msgbox() == 'OK'
+
+    # Test hitting Esc to close.
+    t = KeyPresses([Key.esc])
+    t.start()
+    print('Line', inspect.currentframe().f_lineno)
+    assert easygui.msgbox() is None
+
+    # Test with custom message.
+    t = KeyPresses(' ')
+    t.start()
+    print('Line', inspect.currentframe().f_lineno)
+    assert easygui.msgbox('Message') == 'OK'
+
+    # Test with custom message and title.
+    t = KeyPresses(' ')
+    t.start()
+    print('Line', inspect.currentframe().f_lineno)
+    assert easygui.msgbox('Message', 'Title') == 'OK'
+
+    # Test with custom OK button text.
+    t = KeyPresses(' ')
+    t.start()
+    print('Line', inspect.currentframe().f_lineno)
+    assert easygui.msgbox(ok_button='Button') == 'Button'
+
+    # Test with custom message, title, and OK button text.
+    t = KeyPresses(' ')
+    t.start()
+    print('Line', inspect.currentframe().f_lineno)
+    assert easygui.msgbox('Message', 'Title', ok_button='Button') == 'Button'
+
+    # Test with jpg image.
+    t = KeyPresses(' ')
+    t.start()
+    print('Line', inspect.currentframe().f_lineno)
+    assert easygui.msgbox(image=os.path.join(FOLDER_OF_THIS_FILE, 'pi.jpg')) == 'OK'
+
+    # Test with png image.
+    t = KeyPresses(' ')
+    t.start()
+    print('Line', inspect.currentframe().f_lineno)
+    assert easygui.msgbox(image=os.path.join(FOLDER_OF_THIS_FILE, 'result.png')) == 'OK'
```

Comment:
Can perhaps simplify by parameterizing. Also, we probably want to break this up into separate test cases so the test feedback is a bit easier to reason about and one failed test won't prevent other tests from running.

```suggestion
@pytest.mark.parametrize(
    'args,kwargs,expected',
    (
        ((), {}, 'OK'),  # msgbox with no arguments
        (('Message',), {}, 'OK'),  # with custom message
        (('Message', 'Title'), {}, 'OK'),  # custom message and title
        ((), dict(ok_button='Button'), 'Button'),  # custom button text
        (('Message', 'Title'), dict(ok_button='Button'), 'Button'),  # combo of all three
        ((), dict(image=os.path.join(FOLDER_OF_THIS_FILE, 'pi.jpg')), 'OK'),  # test jpg
        ((), dict(image=os.path.join(FOLDER_OF_THIS_FILE, 'result.png')), 'OK'),  # test png
    ),
)
def test_spacebar_clicks_choice(args, kwargs, expected):
    """
    Test that the spacebar selects a choice.
    Parameterized across several cases, customizing msg, title, etc.
    """

    t = KeyPresses(' ')
    t.start()
    assert easygui.msgbox(*args, **kwargs) == expected
```